cli/cmd/cluster/cluster.go

```diff
@@ -252,3 +253,73 @@ func (c controlplaneUpdater) upgradeComponent(component, namespace string) error
 
 	return nil
 }
+
+// ensureComponent makes sure that given controlplane component is installed and properly configured.
+//
+// Configuration is ensured by forcing a Helm rollback to the latest known version, which should revert
+// all changes done manually to managed resources (restore removed resources etc.).
+//
+//nolint:funlen
+func (c controlplaneUpdater) ensureComponent(component, namespace string) error {
+	actionConfig, err := util.HelmActionConfig(namespace, c.kubeconfig)
+	if err != nil {
+		return fmt.Errorf("initializing Helm action: %w", err)
+	}
+
+	histClient := action.NewHistory(actionConfig)
+	histClient.Max = 1
+
+	history, err := histClient.Run(component)
+	if err != nil && err != driver.ErrReleaseNotFound {
+		return fmt.Errorf("checking for chart history: %w", err)
+	}
+
+	exists := err != driver.ErrReleaseNotFound
+
+	if !exists {
+		fmt.Printf("Controlplane component '%s' is missing, reinstalling...", component)
+
+		helmChart, err := c.getControlplaneChart(component)
+		if err != nil {
+			return fmt.Errorf("loading chart from assets: %w", err)
+		}
+
+		values, err := c.getControlplaneValues(component)
+		if err != nil {
+			return fmt.Errorf("getting chart values from Terraform: %w", err)
+		}
+
+		install := action.NewInstall(actionConfig)
+		install.ReleaseName = component
+		install.Namespace = namespace
+		install.Atomic = true
+		install.CreateNamespace = true
+
+		if _, err := install.Run(helmChart, values); err != nil {
+			fmt.Println("Failed!")
+
+			return fmt.Errorf("installing controlplane component: %w", err)
+		}
+
+		fmt.Println("Done.")
+
+		return nil
+	}
+
+	rollback := action.NewRollback(actionConfig)
```

Comment:
To make sure I understand this:

* If the chart is in the expected state with no user change this does nothing
* If the k8s objects in the chart are manually modified by the user this rolls back to the expected state because the last version Helm is aware of is assumed to be the expected one
* If the chart is manually modified by the user using Helm then we'll be restoring the latest version set by the user but we don't try to handle this case

Is that accurate?

Reply:
>If the chart is in the expected state with no user change this does nothing

Yes, except creating new Secret object with new Release version.

>If the k8s objects in the chart are manually modified by the user this rolls back to the expected state because the last version Helm is aware of is assumed to be the expected one

Correct.

> If the chart is manually modified by the user using Helm then we'll be restoring the latest version set by the user but we don't try to handle this case

Yes. I think this is okay'ish, as it's already an improvement and it's more likely user will tinker things using `kubectl` rather than `helm`. It's not ideal, rather best effort approach. Also, upgrade will later reset things on Helm level back to proper configuration.